internal/component/sql/cleanup.go

```diff
@@ -0,0 +1,241 @@
+	SQLDB                    SQLDB
+	SQLDBWithContext         SQLDBWithContext
+}
+
+type SQLDB interface {
+	Begin(context.Context) (pgx.Tx, error)
+}
+
+type SQLDBWithContext interface {
+	BeginTx(context.Context, *sql.TxOptions) (*sql.Tx, error)
+}
+
+type gc struct {
+	log                      logger.Logger
+	updateLastCleanupQuery   string
+	deleteExpiredValuesQuery string
+	cleanupInterval          time.Duration
+	db                       SQLDB
+	dbWC                     SQLDBWithContext
+
+	closed   atomic.Bool
+	closedCh chan struct{}
+	wg       sync.WaitGroup
+}
+
+func ScheduleGarbageCollector(opts GCOptions) (GarbageCollector, error) {
+	if opts.CleanupInterval <= 0 {
+		return new(gcNoOp), nil
+	}
+
+	if opts.SQLDB == nil && opts.SQLDBWithContext == nil {
+		return nil, fmt.Errorf("either SQLDB or SQLDBWithContext must be provided")
+	}
+	if opts.SQLDB != nil && opts.SQLDBWithContext != nil {
+		return nil, fmt.Errorf("only one of SQLDB or SQLDBWithContext must be provided")
+	}
+
+	gc := &gc{
+		log:                      opts.Logger,
+		updateLastCleanupQuery:   opts.UpdateLastCleanupQuery,
+		deleteExpiredValuesQuery: opts.DeleteExpiredValuesQuery,
+		cleanupInterval:          opts.CleanupInterval,
+		db:                       opts.SQLDB,
+		dbWC:                     opts.SQLDBWithContext,
+		closedCh:                 make(chan struct{}),
+	}
+
+	gc.wg.Add(1)
+	go func() {
+		defer gc.wg.Done()
+		gc.scheduleCleanup()
+	}()
+
+	return gc, nil
+}
+
+func (g *gc) scheduleCleanup() {
+	g.log.Infof("Schedule expired data clean up every %v", g.cleanupInterval)
+
+	ticker := time.NewTicker(g.cleanupInterval)
+	defer ticker.Stop()
+
+	var err error
+	for {
+		select {
+		case <-ticker.C:
+			err = g.CleanupExpired()
+			if err != nil {
+				g.log.Errorf("Error removing expired data: %v", err)
+			}
+		case <-g.closedCh:
+			g.log.Debug("Stopping background cleanup of expired data")
+			return
+		}
+	}
+}
+
+// Exposed for testing.
+func (g *gc) CleanupExpired() error {
+	var (
+		tx   pgx.Tx
+		txwc *sql.Tx
+	)
+
+	// Deletion can take a long time to complete so we have a long background
+	// context. Still catch closing of the GC.
+	ctx, cancel := context.WithTimeout(context.Background(), time.Minute*10)
+	defer cancel()
+
+	g.wg.Add(1)
+	go func() {
+		defer g.wg.Done()
+		defer cancel()
+		select {
+		case <-ctx.Done():
+		case <-g.closedCh:
+		}
+	}()
+
+	var err error
+	if g.db != nil {
+		tx, err = g.db.Begin(ctx)
+		if err != nil {
+			return fmt.Errorf("failed to start transaction: %w", err)
+		}
+		defer tx.Rollback(ctx)
+	} else {
+		txwc, err = g.dbWC.BeginTx(ctx, nil)
+		if err != nil {
+			return fmt.Errorf("failed to start transaction: %w", err)
+		}
+		defer txwc.Rollback()
+	}
+
+	// Check if the last iteration was too recent
+	// This performs an atomic operation, so allows coordination with other daprd processes too
+	// We do this before beginning the transaction
+	canContinue, err := g.updateLastCleanup(ctx, tx, txwc)
```

Comment:
This needs to be called with the DB connection (e.g. `g.db`) and not a transaction, before the transaction lock is acquired.

The reason is that the cleanup operation (so the transaction) can take a long time, so we want to make sure that other daprd processes aren't blocked waiting on the transaction and can hit this "semaphore" before. `updateLastCleanup` is atomic, so there's no risk there.

Reply:
(I changed it already)